server/src/api/muster/muster.controller.ts

```diff
@@ -214,6 +220,115 @@ class MusterController {
 
     res.json(closestMuster ? buildMusterWindow(unit, closestStart, closestEnd, closestMuster) : null);
   }
+
+  /**
+   * This method returns the normalized rate (0 - 1.0) of compliance on a given 
+   * date for a given unit, against  of the unit's muster configs. 
+   * 
+   * @param ReportDateQuery - query param for the date used to check compliance
+   * @param unitName - API param to specify the unit to check for compliance
+   */
+  async getMusterComplianceByDate(req: ApiRequest<{unitName:string}, null, DateQuery>, res: Response) {
+    const outValue = {musterComplianceRate: 0};
+    const reportDate = dateFromString(req.query.timestamp);
+    if (!reportDate) {
+      throw new BadRequestError('Missing reportDate.');
+    }
+    const dayBit = 1 << reportDate.getDay();
+
+    const usersOnRoster = await getUsersOnRosterByDate(req.query.timestamp, req.params.unitName);  
+    const users = usersOnRoster.users;
+    const org = usersOnRoster.org;
+    const userCount = users.length;
+    let musterAvg: number = 0;
+
+    if(usersOnRoster.musterConfig.length == 0 || userCount == 0) {
+      res.json(outValue);
+    }
+
+    for(var i = 0; i < usersOnRoster.musterConfig.length; i++) {
+      const musterConfig = usersOnRoster.musterConfig[i];
+      if(musterConfig.days) {
+        // handle recurring muster...
+        // early out if the reportDate in question is not on a muster day.
+        if((dayBit & musterConfig.days) === dayBit) {
+          const musterTime = reportDate.toISOString().split('T')[0] + " " + musterConfig.startTime; 
+          const musterStart = moment.tz(musterTime, musterConfig.timezone);
+          const musterEnd = moment.tz(musterTime, musterConfig.timezone).add({"minutes": musterConfig.durationMinutes});
+
+          const observationCount = await Observation.createQueryBuilder('observation')
```

Comment:
This code seems to be an exact copy of the one from line 276. I would extract it into a function and call it twice.

Reply:
The calculation for musterStart and musterEnd are different but I will put the query in a function.